File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

let index = 0
const move = [
  [1, 0],
  [-1, 0],
  [0, 1],
  [0, -1],
]
const answer = []
let problem = 1

const bfs = (sx, sy, N, graph) => {
  const queue = [[sx, sy]]
  const distance = Array.from({ length: N }, () =>
    Array.from({ length: N }, () => Number.MAX_SAFE_INTEGER),
  )
  distance[sx][sy] = graph[sx][sy]
  let index = 0

  while (queue.length > index) {
    const [x, y] = queue[index++]
    for (let i = 0; i < 4; i++) {
      const [nx, ny] = [x + move[i][0], y + move[i][1]]
      if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue
      if (distance[nx][ny] > distance[x][y] + graph[nx][ny]) {
        distance[nx][ny] = distance[x][y] + graph[nx][ny]
        queue.push([nx, ny])
      }
    }
  }
  return distance[N - 1][N - 1]
}
while (true) {
  const N = Number(input[index++])
  if (N === 0) break
  const graph = input
    .slice(index, index + N)
    .map((row) => row.split(" ").map(Number))
  index += N
  const dist = bfs(0, 0, N, graph)

  answer.push(`Problem ${problem}: ${dist}`)
  problem+=1
}
console.log(answer.join("\n"))

File: 백준/Gold/4485. 녹색 옷 입은 애가 젤다지？/녹색 옷 입은 애가 젤다지？.py
import sys
from collections import deque
input = sys.stdin.readline

move = [
    (1, 0),
    (-1, 0),
    (0, 1),
    (0, -1),
]
answer = []
problem = 1

def bfs(sx, sy, N, graph):
    queue = deque([(sx, sy)])
    distance = [[float('inf')] * N for _ in range(N)]
    distance[sx][sy] = graph[sx][sy]

    while queue:
        x, y = queue.popleft()
        for dx, dy in move:
            nx, ny = x + dx, y + dy
            if 0 <= nx < N and 0 <= ny < N:
                if distance[nx][ny] > distance[x][y] + graph[nx][ny]:
                    distance[nx][ny] = distance[x][y] + graph[nx][ny]
                    queue.append((nx, ny))
    return distance[N - 1][N - 1]

while True:
    N = int(input().strip())
    if N == 0:
        break
    graph = [list(map(int, input().strip().split())) for _ in range(N)]
    dist = bfs(0, 0, N, graph)
    answer.append(f"Problem {problem}: {dist}")
    problem += 1

print("\n".join(answer))
